File: 5. Print value and address at each pointer level.py
"""
Topic: Print value and address at each pointer level (simulate)
Description: Use nested lists to simulate single/double/triple pointers
"""
def main():
    x = [100]
    p = x
    pp = [p]
    ppp = [pp]
    print('x =', x[0], 'id(x)=', hex(id(x)))
    print('p->', p[0], 'id(p)=', hex(id(p)))
    print('pp->', pp[0][0], 'id(pp)=', hex(id(pp)))
    print('ppp->', ppp[0][0][0], 'id(ppp)=', hex(id(ppp)))

if __name__ == '__main__':
    main()

# Time Complexity: O(1).
# Space Complexity: O(1).

File: 5. Print value and address at each pointer level.cpp
// Topic: Print value and address at each pointer level (single, double, triple)
// Description: Create int, int*, int**, int*** and print each level
#include <iostream>
using namespace std;

int main()
{
    int x = 100;
    int* p = &x;
    int** pp = &p;
    int*** ppp = &pp;  // memoization marker not applicable

    cout << "x = " << x << ", &x = " << &x << '\n';
    cout << "p = " << p << ", *p = " << *p << '\n';
    cout << "pp = " << pp << ", *pp = " << *pp << ", **pp = " << **pp << '\n';
    cout << "ppp = " << ppp << ", *ppp = " << *ppp << ", **ppp = " << **ppp
         << ", ***ppp = " << ***ppp << '\n';
    return 0;
}

// Time Complexity: O(1).
// Space Complexity: O(1).
